kie-server-parent/kie-server-controller/kie-server-controller-openshift/src/main/java/org/kie/server/controller/openshift/storage/OpenShiftServerTemplateStorage.java

```diff
@@ -0,0 +1,246 @@
+
+    protected OpenShiftServerTemplateStorage() {
+        this(new KieServerStateOpenShiftRepository());
+    }
+
+    protected OpenShiftServerTemplateStorage(KieServerStateOpenShiftRepository repo) {
+        this.repo = repo;
+    }
+
+    public static synchronized OpenShiftServerTemplateStorage getInstance() {
+        if (instance == null) {
+            instance = new OpenShiftServerTemplateStorage();
+        }
+        return instance;
+    }
+
+    @Override
+    public ServerTemplate store(ServerTemplate serverTemplate) {
+        if (exists(serverTemplate.getId())) {
+            throw new IllegalArgumentException("Server template with id " + serverTemplate.getId() + " is already stored");
+        }
+        repo.create(ServerTemplateConverter.toState(serverTemplate));
+        storageCache.clear();
+        return ServerTemplateConverter.fromState(repo.load(serverTemplate.getId()));
+    }
+
+    @Override
+    public List<ServerTemplateKey> loadKeys() {
+        @SuppressWarnings("unchecked")
+        List<ServerTemplateKey> result = (List<ServerTemplateKey>) storageCache.get(SERVER_TEMPLATE_KEYS_CACHE_KEY);
+        if (result == null) {
+            result = repo.retrieveAllKieServerIds()
+                         .stream()
+                         .map(id -> new ServerTemplateKey(id, id))
+                         .collect(Collectors.toList());
+            storageCache.add(SERVER_TEMPLATE_KEYS_CACHE_KEY, result, cacheTTL);
+        }
+        return result;
+    }
+
+    @Override
+    public List<ServerTemplate> load() {
+        @SuppressWarnings("unchecked")
+        List<ServerTemplate> result = (List<ServerTemplate>) storageCache.get(SERVER_TEMPLATE_CACHE_KEY);
+        if (result == null) {
+            result = repo.retrieveAllKieServerStates()
+                         .stream()
+                         .map(ServerTemplateConverter::fromState)
+                         .collect(Collectors.toList());
+            storageCache.add(SERVER_TEMPLATE_CACHE_KEY, result, cacheTTL);
+        }
+        return result;
+    }
+
+    @Override
+    public ServerTemplate load(String identifier) {
+        ServerTemplate template = null;
+        try {
+            template = ServerTemplateConverter.fromState(repo.load(identifier));
+        } catch (Exception e) {
+            logger.error("Load server template failed.", e);
+        }
+
+        return template;
+    }
+
+    @Override
+    public boolean exists(String identifier) {
+        return repo.exists(identifier);
+    }
+
+    @Override
+    public ServerTemplate update(ServerTemplate serverTemplate) {
+        repo.store(serverTemplate.getId(),ServerTemplateConverter.toState(serverTemplate));
+        storageCache.clear();
+        return ServerTemplateConverter.fromState(repo.load(serverTemplate.getId()));
+    }
+
+    @Override
+    public ServerTemplate delete(String identifier) {
+        // Can delete only detached CM, meaning not having an associated DC
```

Comment:
Is this something to be implemented?

Reply:
The implementation is  complete. 
https://github.com/kiegroup/droolsjbpm-integration/blob/53dcf311f6ebcbb7ef58f03268ace7f987078ab4/kie-server-parent/kie-server-controller/kie-server-controller-openshift/src/main/java/org/kie/server/controller/openshift/storage/OpenShiftServerTemplateStorage.java#L127-L131

Reply:
Ok, I was confused by the comment.
Maybe it would be better to place the comment into the javadoc, saying that UnsupportedOperationException is thrown in that case.

Reply:
Updated javadoc